From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Fri, 3 Sep 2021 15:50:00 +0200
Subject: [PATCH 1/1] Don't use __declspec(dllexport) for static builds


diff --git a/meson.build b/meson.build
index 1111111..2222222 100644
--- a/meson.build
+++ b/meson.build
@@ -308,7 +308,7 @@ ftoption_h = custom_target('ftoption.h',
 ft2_sources += ftoption_h
 ft2_defines += ['-DFT_CONFIG_OPTIONS_H=<ftoption.h>']
 
-if host_machine.system() == 'windows'
+if get_option('default_library') != 'static' and host_machine.system() == 'windows'
   ft2_defines += ['-DDLL_EXPORT=1']
 endif
 
