From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Behdad Esfahbod <behdad@behdad.org>
Date: Thu, 9 Jul 2026 12:09:01 -0600
Subject: [PATCH 1/1] Ignore -Wunused-template

Fixes https://github.com/harfbuzz/harfbuzz/issues/6059

Upstream-Status: Backport [https://github.com/harfbuzz/harfbuzz/commit/5e887d36722abfef1c02a11ab15a369d7745027e]

diff --git a/src/hb.hh b/src/hb.hh
index 1111111..2222222 100644
--- a/src/hb.hh
+++ b/src/hb.hh
@@ -151,6 +151,7 @@
 #pragma GCC diagnostic ignored "-Wrange-loop-analysis" // https://github.com/harfbuzz/harfbuzz/issues/2834
 #pragma GCC diagnostic ignored "-Wstrict-aliasing"
 #pragma GCC diagnostic ignored "-Wtype-limits"
+#pragma GCC diagnostic ignored "-Wunused-template"
 #pragma GCC diagnostic ignored "-Wc++11-compat" // only gcc raises it
 #endif
