From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Sun, 7 Mar 2021 20:00:00 +0100
Subject: [PATCH 1/1] Avoid __stdcall calling convention on exported functions


diff --git a/include/lcms2.h b/include/lcms2.h
index 1111111..2222222 100644
--- a/include/lcms2.h
+++ b/include/lcms2.h
@@ -233,7 +233,7 @@ typedef int                  cmsBool;
 #        define CMSEXPORT       __stdcall _export
 #        define CMSAPI
 #     else
-#        define CMSEXPORT      __stdcall
+#        define CMSEXPORT
 #        ifdef CMS_DLL_BUILD
 #            define CMSAPI    __declspec(dllexport)
 #        else
