From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Wed, 5 May 2021 21:00:00 +0200
Subject: [PATCH 1/2] Add llvm-mingw ARM/ARM64 targets


diff --git a/vendor/cfg-expr/src/targets/builtins.rs b/vendor/cfg-expr/src/targets/builtins.rs
index 1111111..2222222 100644
--- a/vendor/cfg-expr/src/targets/builtins.rs
+++ b/vendor/cfg-expr/src/targets/builtins.rs
@@ -83,6 +83,16 @@ pub const ALL_BUILTINS: &[TargetInfo<'static>] = &[
         pointer_width: 64,
         endian: Endian::little,
     },
+    TargetInfo {
+        triple: "aarch64-pc-windows-gnu",
+        os: Some(Os::windows),
+        arch: Arch::aarch64,
+        env: Some(Env::gnu),
+        vendor: Some(Vendor::pc),
+        family: Some(Family::windows),
+        pointer_width: 64,
+        endian: Endian::little,
+    },
     TargetInfo {
         triple: "aarch64-pc-windows-msvc",
         os: Some(Os::windows),
@@ -1203,6 +1213,16 @@ pub const ALL_BUILTINS: &[TargetInfo<'static>] = &[
         pointer_width: 32,
         endian: Endian::little,
     },
+    TargetInfo {
+        triple: "thumbv7a-pc-windows-gnu",
+        os: Some(Os::windows),
+        arch: Arch::arm,
+        env: Some(Env::gnu),
+        vendor: Some(Vendor::pc),
+        family: Some(Family::windows),
+        pointer_width: 32,
+        endian: Endian::little,
+    },
     TargetInfo {
         triple: "thumbv7a-pc-windows-msvc",
         os: Some(Os::windows),

From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Mon, 21 Sep 2020 11:00:00 +0200
Subject: [PATCH 2/2] Ensure that compiler_builtins can be successfully built on
 llvm-mingw

Fixes an undefined symbol error while linking the compiler_builtins crate on
llvm-mingw (targeting ARMv7).

Details:
lld-link: error: undefined symbol: __aeabi_idiv
>>> referenced by libcompiler_builtins-<HASH>.rlib(<O_FILE>):(__aeabi_idivmod)

diff --git a/vendor/compiler_builtins/build.rs b/vendor/compiler_builtins/build.rs
index 1111111..2222222 100644
--- a/vendor/compiler_builtins/build.rs
+++ b/vendor/compiler_builtins/build.rs
@@ -319,7 +319,7 @@ mod c {
             }
         }
 
-        if target_arch == "arm" && target_os != "ios" && target_env != "msvc" {
+        if target_arch == "arm" && target_os != "ios" && target_os != "windows" {
             sources.extend(&[
                 ("__aeabi_div0", "arm/aeabi_div0.c"),
                 ("__aeabi_drsub", "arm/aeabi_drsub.c"),
