From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Sat, 6 Mar 2021 19:40:00 +0100
Subject: [PATCH 1/1] Don't export VP8GetCPUInfo in DLLs


diff --git a/src/dsp/dsp.h b/src/dsp/dsp.h
index 1111111..2222222 100644
--- a/src/dsp/dsp.h
+++ b/src/dsp/dsp.h
@@ -225,7 +225,7 @@ typedef enum {
 } CPUFeature;
 // returns true if the CPU supports the feature.
 typedef int (*VP8CPUInfo)(CPUFeature feature);
-WEBP_EXTERN VP8CPUInfo VP8GetCPUInfo;
+extern VP8CPUInfo VP8GetCPUInfo;
 
 //------------------------------------------------------------------------------
 // Init stub generator
