From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Fri, 2 Jul 2021 14:51:52 +0200
Subject: [PATCH 1/1] Also use / as dir separator on Windows


diff --git a/libvips/iofuncs/init.c b/libvips/iofuncs/init.c
index 1111111..2222222 100644
--- a/libvips/iofuncs/init.c
+++ b/libvips/iofuncs/init.c
@@ -1220,7 +1220,7 @@ vips_guess_libdir( const char *argv0, const char *env_name )
 	 */
 	if( strcmp( prefix, VIPS_PREFIX ) == 0 ) 
 		libdir = VIPS_LIBDIR;
-	else if( (suffix = strrchr( VIPS_LIBDIR, G_DIR_SEPARATOR )) )
+	else if( (suffix = strrchr( VIPS_LIBDIR, '/' )) )
 		libdir = g_strdup_printf( "%s%s", prefix, suffix );
 	else
 		libdir = g_strdup_printf( "%s/lib", prefix );
