From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Mon, 26 Aug 2024 13:46:09 +0200
Subject: [PATCH 1/1] text: avoid a `vips_copy()`

+ fix a possible(?) ref leak.

Upstream-Status: Submitted [https://github.com/libvips/libvips/pull/4122]

diff --git a/libvips/create/text.c b/libvips/create/text.c
index 1111111..2222222 100644
--- a/libvips/create/text.c
+++ b/libvips/create/text.c
@@ -392,7 +392,7 @@ vips_text_build(VipsObject *object)
 	VipsObjectClass *class = VIPS_OBJECT_GET_CLASS(object);
 	VipsCreate *create = VIPS_CREATE(object);
 	VipsText *text = (VipsText *) object;
-	VipsImage **t = (VipsImage **) vips_object_local_array(object, 3);
+	VipsImage **t = (VipsImage **) vips_object_local_array(object, 2);
 
 	VipsRect extents;
 	VipsImage *image;
@@ -471,20 +471,21 @@ vips_text_build(VipsObject *object)
 		return -1;
 	}
 
-	image = t[0] = vips_image_new_memory();
-	vips_image_init_fields(image,
+	t[0] = vips_image_new_memory();
+	vips_image_init_fields(t[0],
 		extents.width, extents.height, 4,
 		VIPS_FORMAT_UCHAR, VIPS_CODING_NONE,
 		VIPS_INTERPRETATION_sRGB,
 		text->dpi / 25.4, text->dpi / 25.4);
-	image->Xoffset = extents.left;
-	image->Yoffset = extents.top;
+	t[0]->Xoffset = extents.left;
+	t[0]->Yoffset = extents.top;
 
-	if (vips_image_pipelinev(image, VIPS_DEMAND_STYLE_ANY, NULL) ||
-		vips_image_write_prepare(image)) {
+	if (vips_image_pipelinev(t[0], VIPS_DEMAND_STYLE_ANY, NULL) ||
+		vips_image_write_prepare(t[0])) {
 		g_mutex_unlock(vips_text_lock);
 		return -1;
 	}
+	image = t[0];
 
 	surface = cairo_image_surface_create_for_data(
 		VIPS_IMAGE_ADDR(image, 0, 0),
@@ -527,12 +528,11 @@ vips_text_build(VipsObject *object)
 	else {
 		/* We just want the alpha channel.
 		 */
-		if (vips_extract_band(image, &t[1], 3, NULL) ||
-			vips_copy(t[1], &t[2],
-				"interpretation", VIPS_INTERPRETATION_MULTIBAND,
-				NULL))
+		if (vips_extract_band(image, &t[1], 3, NULL))
 			return -1;
-		image = t[2];
+		image = t[1];
+
+		image->Type = VIPS_INTERPRETATION_MULTIBAND;
 	}
 
 	if (vips_image_write(image, create->out))
