From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martin Bruse <zondolfin@gmail.com>
Date: Mon, 8 Jul 2024 11:46:19 +0200
Subject: [PATCH 1/1] Made it possible to force usage of system JPEG libraries.

Upstream-Status: Submitted [https://github.com/libjxl/libjxl/pull/3704]

diff --git a/lib/jpegli.cmake b/lib/jpegli.cmake
index 1111111..2222222 100644
--- a/lib/jpegli.cmake
+++ b/lib/jpegli.cmake
@@ -24,12 +24,30 @@ elseif(JPEGLI_LIBJPEG_LIBRARY_SOVERSION STREQUAL "8")
   set(JPEG_LIB_VERSION 80)
 endif()
 
-configure_file(
-  ../third_party/libjpeg-turbo/jconfig.h.in include/jpegli/jconfig.h)
-configure_file(
-  ../third_party/libjpeg-turbo/jpeglib.h include/jpegli/jpeglib.h COPYONLY)
-configure_file(
-  ../third_party/libjpeg-turbo/jmorecfg.h include/jpegli/jmorecfg.h COPYONLY)
+# Force system dependencies.
+set(JPEGXL_FORCE_SYSTEM_JPEG_TURBO false CACHE BOOL
+    "Force using system installed jpegturbo instead of third_party/libjpeg-turbo source.")
+
+# libjpeg-turbo
+find_package(JPEG)
+
+if (NOT EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/libjpeg-turbo/jpeglib.h" OR
+    JPEGXL_FORCE_SYSTEM_JPEG_TURBO)
+  if (NOT JPEG_FOUND)
+    message(FATAL_ERROR
+         "JPEG not found, install libjpeg62-turbo and libjpeg62-turbo-dev or"
+         " download libjpeg-turbo source to third_party/libjpeg-turbo from"
+         " https://github.com/libjpeg-turbo/libjpeg-turbo. You can use"
+         " ${PROJECT_SOURCE_DIR}/deps.sh to download this dependency.")
+  endif()
+else()
+  configure_file(
+    ../third_party/libjpeg-turbo/jconfig.h.in include/jpegli/jconfig.h)
+  configure_file(
+    ../third_party/libjpeg-turbo/jpeglib.h include/jpegli/jpeglib.h COPYONLY)
+  configure_file(
+    ../third_party/libjpeg-turbo/jmorecfg.h include/jpegli/jmorecfg.h COPYONLY)
+endif()
 
 add_library(jpegli-static STATIC EXCLUDE_FROM_ALL "${JPEGXL_INTERNAL_JPEGLI_SOURCES}")
 target_compile_options(jpegli-static PRIVATE "${JPEGXL_INTERNAL_FLAGS}")
@@ -48,58 +66,57 @@ target_link_libraries(jpegli-static PUBLIC ${JPEGLI_INTERNAL_LIBS})
 # Tests for jpegli-static
 #
 
-find_package(JPEG)
 if(JPEG_FOUND AND BUILD_TESTING)
-# TODO(eustas): merge into jxl_tests.cmake?
+  # TODO(eustas): merge into jxl_tests.cmake?
 
-add_library(jpegli_libjpeg_util-obj OBJECT
-  ${JPEGXL_INTERNAL_JPEGLI_LIBJPEG_HELPER_FILES}
-)
-target_include_directories(jpegli_libjpeg_util-obj PRIVATE
-  "${PROJECT_SOURCE_DIR}"
-  "${JPEG_INCLUDE_DIRS}"
-)
-target_compile_options(jpegli_libjpeg_util-obj PRIVATE
-  "${JPEGXL_INTERNAL_FLAGS}" "${JPEGXL_COVERAGE_FLAGS}")
-
-# Individual test binaries:
-file(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/tests)
-foreach (TESTFILE IN LISTS JPEGXL_INTERNAL_JPEGLI_TESTS)
-  # The TESTNAME is the name without the extension or directory.
-  get_filename_component(TESTNAME ${TESTFILE} NAME_WE)
-  add_executable(${TESTNAME} ${TESTFILE}
-    $<TARGET_OBJECTS:jpegli_libjpeg_util-obj>
-    ${JPEGXL_INTERNAL_JPEGLI_TESTLIB_FILES}
+  add_library(jpegli_libjpeg_util-obj OBJECT
+    ${JPEGXL_INTERNAL_JPEGLI_LIBJPEG_HELPER_FILES}
   )
-  target_compile_options(${TESTNAME} PRIVATE
-    ${JPEGXL_INTERNAL_FLAGS}
-    # Add coverage flags to the test binary so code in the private headers of
-    # the library is also instrumented when running tests that execute it.
-    ${JPEGXL_COVERAGE_FLAGS}
-  )
-  target_compile_definitions(${TESTNAME} PRIVATE
-    -DTEST_DATA_PATH="${JPEGXL_TEST_DATA_PATH}")
-  target_include_directories(${TESTNAME} PRIVATE
+  target_include_directories(jpegli_libjpeg_util-obj PRIVATE
     "${PROJECT_SOURCE_DIR}"
-    "${CMAKE_CURRENT_SOURCE_DIR}/include"
-    "${CMAKE_CURRENT_BINARY_DIR}/include"
-  )
-  target_link_libraries(${TESTNAME}
-    hwy
-    jpegli-static
-    GTest::GTest
-    GTest::Main
-    ${JPEG_LIBRARIES}
+    "${JPEG_INCLUDE_DIRS}"
   )
-  set_target_properties(${TESTNAME} PROPERTIES LINK_FLAGS "${JPEGXL_COVERAGE_LINK_FLAGS}")
-  # Output test targets in the test directory.
-  set_target_properties(${TESTNAME} PROPERTIES PREFIX "tests/")
-  if (WIN32 AND CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
-    set_target_properties(${TESTNAME} PROPERTIES COMPILE_FLAGS "-Wno-error")
-  endif ()
-  # 240 seconds because some build types (e.g. coverage) can be quite slow.
-  gtest_discover_tests(${TESTNAME} DISCOVERY_TIMEOUT 240)
-endforeach ()
+  target_compile_options(jpegli_libjpeg_util-obj PRIVATE
+    "${JPEGXL_INTERNAL_FLAGS}" "${JPEGXL_COVERAGE_FLAGS}")
+
+  # Individual test binaries:
+  file(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/tests)
+  foreach (TESTFILE IN LISTS JPEGXL_INTERNAL_JPEGLI_TESTS)
+    # The TESTNAME is the name without the extension or directory.
+    get_filename_component(TESTNAME ${TESTFILE} NAME_WE)
+    add_executable(${TESTNAME} ${TESTFILE}
+      $<TARGET_OBJECTS:jpegli_libjpeg_util-obj>
+      ${JPEGXL_INTERNAL_JPEGLI_TESTLIB_FILES}
+    )
+    target_compile_options(${TESTNAME} PRIVATE
+      ${JPEGXL_INTERNAL_FLAGS}
+      # Add coverage flags to the test binary so code in the private headers of
+      # the library is also instrumented when running tests that execute it.
+      ${JPEGXL_COVERAGE_FLAGS}
+    )
+    target_compile_definitions(${TESTNAME} PRIVATE
+      -DTEST_DATA_PATH="${JPEGXL_TEST_DATA_PATH}")
+    target_include_directories(${TESTNAME} PRIVATE
+      "${PROJECT_SOURCE_DIR}"
+      "${CMAKE_CURRENT_SOURCE_DIR}/include"
+      "${CMAKE_CURRENT_BINARY_DIR}/include"
+    )
+    target_link_libraries(${TESTNAME}
+      hwy
+      jpegli-static
+      GTest::GTest
+      GTest::Main
+      ${JPEG_LIBRARIES}
+    )
+    set_target_properties(${TESTNAME} PROPERTIES LINK_FLAGS "${JPEGXL_COVERAGE_LINK_FLAGS}")
+    # Output test targets in the test directory.
+    set_target_properties(${TESTNAME} PROPERTIES PREFIX "tests/")
+    if (WIN32 AND CMAKE_CXX_COMPILER_ID STREQUAL "Clang")
+      set_target_properties(${TESTNAME} PROPERTIES COMPILE_FLAGS "-Wno-error")
+    endif ()
+    # 240 seconds because some build types (e.g. coverage) can be quite slow.
+    gtest_discover_tests(${TESTNAME} DISCOVERY_TIMEOUT 240)
+  endforeach ()
 endif()
 
 #
@@ -107,51 +124,51 @@ endif()
 #
 
 if (JPEGXL_ENABLE_JPEGLI_LIBJPEG AND NOT APPLE AND NOT WIN32 AND NOT EMSCRIPTEN)
-add_library(jpegli-libjpeg-obj OBJECT "${JPEGXL_INTERNAL_JPEGLI_WRAPPER_SOURCES}")
-target_compile_options(jpegli-libjpeg-obj PRIVATE ${JPEGXL_INTERNAL_FLAGS})
-target_compile_options(jpegli-libjpeg-obj PUBLIC ${JPEGXL_COVERAGE_FLAGS})
-set_property(TARGET jpegli-libjpeg-obj PROPERTY POSITION_INDEPENDENT_CODE ON)
-target_include_directories(jpegli-libjpeg-obj PRIVATE
-  "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}>"
-  "$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include/jpegli>"
-)
-target_compile_definitions(jpegli-libjpeg-obj PUBLIC
-  ${JPEGLI_LIBJPEG_OBJ_COMPILE_DEFINITIONS}
-)
-set(JPEGLI_LIBJPEG_INTERNAL_OBJECTS $<TARGET_OBJECTS:jpegli-libjpeg-obj>)
-
-file(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/jpegli)
-add_library(jpeg SHARED ${JPEGLI_LIBJPEG_INTERNAL_OBJECTS})
-target_link_libraries(jpeg PUBLIC ${JPEGXL_COVERAGE_FLAGS})
-target_link_libraries(jpeg PRIVATE jpegli-static)
-set_target_properties(jpeg PROPERTIES
-  VERSION ${JPEGLI_LIBJPEG_LIBRARY_VERSION}
-  SOVERSION ${JPEGLI_LIBJPEG_LIBRARY_SOVERSION}
-  LIBRARY_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/jpegli"
-  RUNTIME_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/jpegli")
-
-# Add a jpeg.version file as a version script to tag symbols with the
-# appropriate version number.
-set_target_properties(jpeg PROPERTIES
-  LINK_DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/jpegli/jpeg.version.${JPEGLI_LIBJPEG_LIBRARY_SOVERSION})
-set_property(TARGET jpeg APPEND_STRING PROPERTY
-  LINK_FLAGS " -Wl,--version-script=${CMAKE_CURRENT_SOURCE_DIR}/jpegli/jpeg.version.${JPEGLI_LIBJPEG_LIBRARY_SOVERSION}")
-
-if (JPEGXL_INSTALL_JPEGLI_LIBJPEG)
-  install(TARGETS jpeg
-    RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
-    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
-    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
-  install(
-    DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/include/jpegli/"
-    DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}")
-endif()
-
-# This hides the default visibility symbols from static libraries bundled into
-# the shared library. In particular this prevents exposing symbols from hwy
-# in the shared library.
-if(LINKER_SUPPORT_EXCLUDE_LIBS)
+  add_library(jpegli-libjpeg-obj OBJECT "${JPEGXL_INTERNAL_JPEGLI_WRAPPER_SOURCES}")
+  target_compile_options(jpegli-libjpeg-obj PRIVATE ${JPEGXL_INTERNAL_FLAGS})
+  target_compile_options(jpegli-libjpeg-obj PUBLIC ${JPEGXL_COVERAGE_FLAGS})
+  set_property(TARGET jpegli-libjpeg-obj PROPERTY POSITION_INDEPENDENT_CODE ON)
+  target_include_directories(jpegli-libjpeg-obj PRIVATE
+    "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}>"
+    "$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include/jpegli>"
+  )
+  target_compile_definitions(jpegli-libjpeg-obj PUBLIC
+    ${JPEGLI_LIBJPEG_OBJ_COMPILE_DEFINITIONS}
+  )
+  set(JPEGLI_LIBJPEG_INTERNAL_OBJECTS $<TARGET_OBJECTS:jpegli-libjpeg-obj>)
+
+  file(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/jpegli)
+  add_library(jpeg SHARED ${JPEGLI_LIBJPEG_INTERNAL_OBJECTS})
+  target_link_libraries(jpeg PUBLIC ${JPEGXL_COVERAGE_FLAGS})
+  target_link_libraries(jpeg PRIVATE jpegli-static)
+  set_target_properties(jpeg PROPERTIES
+    VERSION ${JPEGLI_LIBJPEG_LIBRARY_VERSION}
+    SOVERSION ${JPEGLI_LIBJPEG_LIBRARY_SOVERSION}
+    LIBRARY_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/jpegli"
+    RUNTIME_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/jpegli")
+
+  # Add a jpeg.version file as a version script to tag symbols with the
+  # appropriate version number.
+  set_target_properties(jpeg PROPERTIES
+    LINK_DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/jpegli/jpeg.version.${JPEGLI_LIBJPEG_LIBRARY_SOVERSION})
   set_property(TARGET jpeg APPEND_STRING PROPERTY
-    LINK_FLAGS " ${LINKER_EXCLUDE_LIBS_FLAG}")
-endif()
+    LINK_FLAGS " -Wl,--version-script=${CMAKE_CURRENT_SOURCE_DIR}/jpegli/jpeg.version.${JPEGLI_LIBJPEG_LIBRARY_SOVERSION}")
+
+  if (JPEGXL_INSTALL_JPEGLI_LIBJPEG)
+    install(TARGETS jpeg
+      RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+      LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+      ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
+    install(
+      DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/include/jpegli/"
+      DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}")
+  endif()
+
+  # This hides the default visibility symbols from static libraries bundled into
+  # the shared library. In particular this prevents exposing symbols from hwy
+  # in the shared library.
+  if(LINKER_SUPPORT_EXCLUDE_LIBS)
+    set_property(TARGET jpeg APPEND_STRING PROPERTY
+      LINK_FLAGS " ${LINKER_EXCLUDE_LIBS_FLAG}")
+  endif()
 endif()

From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Thu, 7 Mar 2024 18:28:49 +0100
Subject: [PATCH 2/2] Build fixes for Windows

Upstream-Status: Pending

diff --git a/lib/jpegli.cmake b/lib/jpegli.cmake
index 1111111..2222222 100644
--- a/lib/jpegli.cmake
+++ b/lib/jpegli.cmake
@@ -47,23 +47,26 @@ else()
     ../third_party/libjpeg-turbo/jpeglib.h include/jpegli/jpeglib.h COPYONLY)
   configure_file(
     ../third_party/libjpeg-turbo/jmorecfg.h include/jpegli/jmorecfg.h COPYONLY)
+  configure_file(
+    ../third_party/libjpeg-turbo/jerror.h include/jpegli/jerror.h COPYONLY)
 endif()
 
-add_library(jpegli-static STATIC EXCLUDE_FROM_ALL "${JPEGXL_INTERNAL_JPEGLI_SOURCES}")
-target_compile_options(jpegli-static PRIVATE "${JPEGXL_INTERNAL_FLAGS}")
-target_compile_options(jpegli-static PUBLIC ${JPEGXL_COVERAGE_FLAGS})
-set_property(TARGET jpegli-static PROPERTY POSITION_INDEPENDENT_CODE ON)
-target_include_directories(jpegli-static PRIVATE
-  "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}>"
-  "${JXL_HWY_INCLUDE_DIRS}"
+add_library(jpegli-obj OBJECT "${JPEGXL_INTERNAL_JPEGLI_SOURCES}")
+set_target_properties(jpegli-obj PROPERTIES POSITION_INDEPENDENT_CODE ON)
+target_compile_options(jpegli-obj
+  PUBLIC "${JPEGXL_COVERAGE_FLAGS}"
+  PRIVATE "${JPEGXL_INTERNAL_FLAGS}"
 )
-target_include_directories(jpegli-static PUBLIC
-  "$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include/jpegli>"
+target_include_directories(jpegli-obj
+  PUBLIC
+    "$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include/jpegli>"
+  PRIVATE
+    "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}>"
+    "${JXL_HWY_INCLUDE_DIRS}"
 )
-target_link_libraries(jpegli-static PUBLIC ${JPEGLI_INTERNAL_LIBS})
 
 #
-# Tests for jpegli-static
+# Tests for jpegli
 #
 
 if(JPEG_FOUND AND BUILD_TESTING)
@@ -85,6 +88,7 @@ if(JPEG_FOUND AND BUILD_TESTING)
     # The TESTNAME is the name without the extension or directory.
     get_filename_component(TESTNAME ${TESTFILE} NAME_WE)
     add_executable(${TESTNAME} ${TESTFILE}
+      $<TARGET_OBJECTS:jpegli-obj>
       $<TARGET_OBJECTS:jpegli_libjpeg_util-obj>
       ${JPEGXL_INTERNAL_JPEGLI_TESTLIB_FILES}
     )
@@ -102,11 +106,12 @@ if(JPEG_FOUND AND BUILD_TESTING)
       "${CMAKE_CURRENT_BINARY_DIR}/include"
     )
     target_link_libraries(${TESTNAME}
-      hwy
-      jpegli-static
-      GTest::GTest
-      GTest::Main
-      ${JPEG_LIBRARIES}
+      PUBLIC
+        GTest::GTest
+        GTest::Main
+        ${JPEG_LIBRARIES}
+      PRIVATE
+        ${JPEGLI_INTERNAL_LIBS}
     )
     set_target_properties(${TESTNAME} PROPERTIES LINK_FLAGS "${JPEGXL_COVERAGE_LINK_FLAGS}")
     # Output test targets in the test directory.
@@ -120,48 +125,91 @@ if(JPEG_FOUND AND BUILD_TESTING)
 endif()
 
 #
-# Build libjpeg.so that links to libjpeg-static
+# Build libjpeg.{so,a,dll,lib} that uses the objects and properties of jpegli-obj
 #
 
-if (JPEGXL_ENABLE_JPEGLI_LIBJPEG AND NOT APPLE AND NOT WIN32 AND NOT EMSCRIPTEN)
-  add_library(jpegli-libjpeg-obj OBJECT "${JPEGXL_INTERNAL_JPEGLI_WRAPPER_SOURCES}")
-  target_compile_options(jpegli-libjpeg-obj PRIVATE ${JPEGXL_INTERNAL_FLAGS})
-  target_compile_options(jpegli-libjpeg-obj PUBLIC ${JPEGXL_COVERAGE_FLAGS})
-  set_property(TARGET jpegli-libjpeg-obj PROPERTY POSITION_INDEPENDENT_CODE ON)
-  target_include_directories(jpegli-libjpeg-obj PRIVATE
+if (JPEGXL_ENABLE_JPEGLI_LIBJPEG)
+  add_library(jpeg
+    $<TARGET_OBJECTS:jpegli-obj>
+    "${JPEGXL_INTERNAL_JPEGLI_WRAPPER_SOURCES}"
+  )
+  target_link_libraries(jpeg
+    PUBLIC ${JPEGXL_COVERAGE_FLAGS}
+    PRIVATE ${JPEGLI_INTERNAL_LIBS}
+  )
+  target_compile_options(jpeg PRIVATE ${JPEGXL_INTERNAL_FLAGS})
+  target_include_directories(jpeg PRIVATE
     "$<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}>"
     "$<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/include/jpegli>"
   )
-  target_compile_definitions(jpegli-libjpeg-obj PUBLIC
-    ${JPEGLI_LIBJPEG_OBJ_COMPILE_DEFINITIONS}
-  )
-  set(JPEGLI_LIBJPEG_INTERNAL_OBJECTS $<TARGET_OBJECTS:jpegli-libjpeg-obj>)
-
-  file(MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/jpegli)
-  add_library(jpeg SHARED ${JPEGLI_LIBJPEG_INTERNAL_OBJECTS})
-  target_link_libraries(jpeg PUBLIC ${JPEGXL_COVERAGE_FLAGS})
-  target_link_libraries(jpeg PRIVATE jpegli-static)
   set_target_properties(jpeg PROPERTIES
+    POSITION_INDEPENDENT_CODE ON
     VERSION ${JPEGLI_LIBJPEG_LIBRARY_VERSION}
     SOVERSION ${JPEGLI_LIBJPEG_LIBRARY_SOVERSION}
     LIBRARY_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/jpegli"
     RUNTIME_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/jpegli")
 
-  # Add a jpeg.version file as a version script to tag symbols with the
-  # appropriate version number.
-  set_target_properties(jpeg PROPERTIES
-    LINK_DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/jpegli/jpeg.version.${JPEGLI_LIBJPEG_LIBRARY_SOVERSION})
-  set_property(TARGET jpeg APPEND_STRING PROPERTY
-    LINK_FLAGS " -Wl,--version-script=${CMAKE_CURRENT_SOURCE_DIR}/jpegli/jpeg.version.${JPEGLI_LIBJPEG_LIBRARY_SOVERSION}")
+  if (APPLE)
+    set_property(TARGET jpeg APPEND_STRING PROPERTY
+      LINK_FLAGS "-Wl,-exported_symbol,_jpeg* -Wl,-exported_symbol,_tj*")
+  elseif (WIN32)
+    # Create jpeg<SOVERSION>.dll or libjpeg-<SOVERSION>.dll when building a
+    # shared library on Windows.
+    if (MSVC)
+      set_target_properties(jpeg PROPERTIES
+        RUNTIME_OUTPUT_NAME jpeg${JPEGLI_LIBJPEG_LIBRARY_SOVERSION})
+    elseif (MINGW)
+      set_target_properties(jpeg PROPERTIES
+        RUNTIME_OUTPUT_NAME jpeg-${JPEGLI_LIBJPEG_LIBRARY_SOVERSION})
+    endif()
+  else()
+    # Add a jpeg.version file as a version script to tag symbols with the
+    # appropriate version number.
+    set_target_properties(jpeg PROPERTIES
+      LINK_DEPENDS ${CMAKE_CURRENT_SOURCE_DIR}/jpegli/jpeg.version.${JPEGLI_LIBJPEG_LIBRARY_SOVERSION})
+    set_property(TARGET jpeg APPEND_STRING PROPERTY
+      LINK_FLAGS " -Wl,--version-script=${CMAKE_CURRENT_SOURCE_DIR}/jpegli/jpeg.version.${JPEGLI_LIBJPEG_LIBRARY_SOVERSION}")
+  endif()  # APPLE
 
   if (JPEGXL_INSTALL_JPEGLI_LIBJPEG)
     install(TARGETS jpeg
       RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
       LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
       ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
-    install(
-      DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/include/jpegli/"
-      DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}")
+
+    if (NOT JPEGXL_FORCE_SYSTEM_JPEG_TURBO)
+      install(
+        DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/include/jpegli/"
+        DESTINATION "${CMAKE_INSTALL_INCLUDEDIR}")
+    endif()
+
+    ### Add a pkg-config file for libjpeg.
+
+    # Allow adding prefix if CMAKE_INSTALL_INCLUDEDIR not absolute.
+    if (IS_ABSOLUTE "${CMAKE_INSTALL_INCLUDEDIR}")
+      set(PKGCONFIG_TARGET_INCLUDES "${CMAKE_INSTALL_INCLUDEDIR}")
+    else()
+      set(PKGCONFIG_TARGET_INCLUDES "\${prefix}/${CMAKE_INSTALL_INCLUDEDIR}")
+    endif()
+    # Allow adding prefix if CMAKE_INSTALL_LIBDIR not absolute.
+    if (IS_ABSOLUTE "${CMAKE_INSTALL_LIBDIR}")
+      set(PKGCONFIG_TARGET_LIBS "${CMAKE_INSTALL_LIBDIR}")
+    else()
+      set(PKGCONFIG_TARGET_LIBS "\${exec_prefix}/${CMAKE_INSTALL_LIBDIR}")
+    endif()
+
+    if (BUILD_SHARED_LIBS)
+      set(JPEGLI_REQUIRES_TYPE "Requires.private")
+      set(JPEGLI_PRIVATE_LIBS "-lm ${PKGCONFIG_CXX_LIB}")
+    else()
+      set(JPEGLI_REQUIRES_TYPE "Requires")
+      set(JPEGLI_PUBLIC_LIBS "-lm ${PKGCONFIG_CXX_LIB}")
+    endif()
+
+    configure_file("${CMAKE_CURRENT_SOURCE_DIR}/jpegli/libjpeg.pc.in"
+                   "libjpeg.pc" @ONLY)
+    install(FILES "${CMAKE_CURRENT_BINARY_DIR}/libjpeg.pc"
+            DESTINATION "${CMAKE_INSTALL_LIBDIR}/pkgconfig")
   endif()
 
   # This hides the default visibility symbols from static libraries bundled into
diff --git a/lib/jpegli/libjpeg.pc.in b/lib/jpegli/libjpeg.pc.in
new file mode 100644
index 0000000..1111111
--- /dev/null
+++ b/lib/jpegli/libjpeg.pc.in
@@ -0,0 +1,12 @@
+prefix=@CMAKE_INSTALL_PREFIX@
+exec_prefix=${prefix}
+libdir=@PKGCONFIG_TARGET_LIBS@
+includedir=@PKGCONFIG_TARGET_INCLUDES@
+
+Name: libjpeg
+Description: Improved JPEG encoder and decoder implementation
+Version: @JPEGXL_LIBRARY_VERSION@
+@JPEGLI_REQUIRES_TYPE@: libhwy
+Libs: -L${libdir} -ljpeg @JPEGLI_PUBLIC_LIBS@
+Libs.private: @JPEGLI_PRIVATE_LIBS@
+Cflags: -I${includedir}
