From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kleis Auke Wolthuizen <github@kleisauke.nl>
Date: Mon, 9 Mar 2020 14:00:00 +0100
Subject: [PATCH 1/1] Fix CPPFLAGS for libarm32 dinput/dinput8


diff --git a/mingw-w64-crt/Makefile.am b/mingw-w64-crt/Makefile.am
index 1111111..2222222 100644
--- a/mingw-w64-crt/Makefile.am
+++ b/mingw-w64-crt/Makefile.am
@@ -1326,12 +1326,12 @@ libarm32_libshell32_a_CPPFLAGS=$(CPPFLAGSARM32) $(sysincludes)
 
 libarm32_LIBRARIES += libarm32/libdinput.a
 libarm32_libdinput_a_SOURCES = $(src_libdinput)
-libarm32_libdinput_a_CPPFLAGS=$(CPPFLAGS64) $(sysincludes)
+libarm32_libdinput_a_CPPFLAGS=$(CPPFLAGSARM32) $(sysincludes)
 libarm32_libdinput_a_AR = $(DTDEFARM32) $(top_srcdir)/lib-common/dinput.def && $(AR) $(ARFLAGS)
 
 libarm32_LIBRARIES += libarm32/libdinput8.a
 libarm32_libdinput8_a_SOURCES = $(src_libdinput8)
-libarm32_libdinput8_a_CPPFLAGS=$(CPPFLAGS64) $(sysincludes)
+libarm32_libdinput8_a_CPPFLAGS=$(CPPFLAGSARM32) $(sysincludes)
 libarm32_libdinput8_a_AR = $(DTDEFARM32) $(top_srcdir)/lib-common/dinput8.def && $(AR) $(ARFLAGS)
 
 libarm32_LIBRARIES += libarm32/libdmoguids.a
diff --git a/mingw-w64-crt/Makefile.in b/mingw-w64-crt/Makefile.in
index 1111111..2222222 100644
--- a/mingw-w64-crt/Makefile.in
+++ b/mingw-w64-crt/Makefile.in
@@ -12656,10 +12656,10 @@ src_intrincsarm64 =
 @LIBARM32_TRUE@libarm32_libshell32_a_AR = $(DTDEFARM32) $(top_srcdir)/lib-common/shell32.def && $(AR) $(ARFLAGS)
 @LIBARM32_TRUE@libarm32_libshell32_a_CPPFLAGS = $(CPPFLAGSARM32) $(sysincludes)
 @LIBARM32_TRUE@libarm32_libdinput_a_SOURCES = $(src_libdinput)
-@LIBARM32_TRUE@libarm32_libdinput_a_CPPFLAGS = $(CPPFLAGS64) $(sysincludes)
+@LIBARM32_TRUE@libarm32_libdinput_a_CPPFLAGS = $(CPPFLAGSARM32) $(sysincludes)
 @LIBARM32_TRUE@libarm32_libdinput_a_AR = $(DTDEFARM32) $(top_srcdir)/lib-common/dinput.def && $(AR) $(ARFLAGS)
 @LIBARM32_TRUE@libarm32_libdinput8_a_SOURCES = $(src_libdinput8)
-@LIBARM32_TRUE@libarm32_libdinput8_a_CPPFLAGS = $(CPPFLAGS64) $(sysincludes)
+@LIBARM32_TRUE@libarm32_libdinput8_a_CPPFLAGS = $(CPPFLAGSARM32) $(sysincludes)
 @LIBARM32_TRUE@libarm32_libdinput8_a_AR = $(DTDEFARM32) $(top_srcdir)/lib-common/dinput8.def && $(AR) $(ARFLAGS)
 @LIBARM32_TRUE@libarm32_libdmoguids_a_SOURCES = $(src_libdmoguids)
 @LIBARM32_TRUE@libarm32_libdmoguids_a_CPPFLAGS = $(CPPFLAGSARM32) $(sysincludes)
